cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillController.java

```diff
@@ -0,0 +1,314 @@
+        state = report.getYarnApplicationState();
+        LOG.debug("Yarn application status for {} {}: {}", appName, appId, state);
+        TimeUnit.SECONDS.sleep(1);
+        stopWatch.split();
+      }
+      LOG.info("Yarn application {} {} is in state {}", appName, appId, state);
+      if (state != YarnApplicationState.RUNNING) {
+        LOG.info("Yarn application {} {} is not in running state. Shutting down controller.",
+                 appName, appId, Constants.APPLICATION_MAX_START_SECONDS);
+        forceShutDown();
+      } else {
+        try {
+          URL resourceUrl = URI.create(String.format("http://%s:%d", report.getHost(), report.getRpcPort()))
+            .resolve(TrackerService.PATH).toURL();
+          resourcesClient = new ResourceReportClient(resourceUrl);
+        } catch (IOException e) {
+          resourcesClient = null;
+        }
+      }
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  protected synchronized void doShutDown() {
+    if (processController == null) {
+      LOG.warn("No process controller for application that is not submitted.");
+      return;
+    }
+
+    // Stop polling if it is running.
+    stopPollStatus();
+
+    // Wait for the stop message being processed
+    try {
+      Uninterruptibles.getUninterruptibly(getStopMessageFuture(),
+                                          Constants.APPLICATION_MAX_STOP_SECONDS, TimeUnit.SECONDS);
+    } catch (Exception e) {
+      LOG.error("Failed to wait for stop message being processed.", e);
+      // Kill the application through yarn
+      kill();
+    }
+
+    // Poll application status from yarn
+    FinalApplicationStatus finalStatus = null;
+    try {
+      StopWatch stopWatch = new StopWatch();
+      stopWatch.start();
+      stopWatch.split();
+      long maxTime = TimeUnit.MILLISECONDS.convert(Constants.APPLICATION_MAX_STOP_SECONDS, TimeUnit.SECONDS);
+
+      YarnApplicationReport report = processController.getReport();
+      finalStatus = report.getFinalApplicationStatus();
+      ApplicationId appId = report.getApplicationId();
+      while (finalStatus == FinalApplicationStatus.UNDEFINED && stopWatch.getSplitTime() < maxTime) {
+        LOG.debug("Yarn application final status for {} {}: {}", appName, appId, finalStatus);
+        TimeUnit.SECONDS.sleep(1);
+        stopWatch.split();
+        finalStatus = processController.getReport().getFinalApplicationStatus();
+      }
+      LOG.debug("Yarn application {} {} completed with status {}", appName, appId, finalStatus);
+
+      // Application not finished after max stop time, kill the application
+      if (finalStatus == FinalApplicationStatus.UNDEFINED) {
+        kill();
+      }
+    } catch (Exception e) {
+      LOG.warn("Exception while waiting for application report: {}", e.getMessage(), e);
+      kill();
+    }
+
+    super.doShutDown();
+
+    if (finalStatus == FinalApplicationStatus.FAILED) {
```

Comment:
Line 188-192 are newly added. Also the `finalStatus` variable was pulled out of the try block.